SSR + CSR 방식으로 리스트 출력 및 갱신하기

42KIM opened the pull request:
## <SSR + CSR 방식으로 데이터 출력 및 갱신하기>

### 마이페이지 '진행 중인 스터디'로 테스트 진행

+ 최초 렌더링은 서버사이드에서, 추가적인 렌더링은 클라이언트 사이드에서 이루어지는 시나리오이다.
+ DB 부분은 서버 사이드 렌더링과 같은 환경
+ 마찬가지로 사용자 식별도 구현하지 않은 상태
+ 첫번째 시나리오와 비교했을 때 각각의 장점과 단점이 있는 듯(속도, 보안, 통신 비용 등등). 최대한 면밀히 비교해봐야 할 것 같음


### 진행 흐름

1. 사용자가 최초 마이페이지 방문 시(```/mypage```), 서버는 ```routes/mypage.js```에서 DB로부터 4개(추후 화면 css에 맞게 개수 변경)의 '진행 중인 스터디' 데이터를 가져와 ```mypage.ejs```에 전달한다. 서버는 해당 데이터가 적용된 ```mypage.ejs``` 파일을 렌더링한다. (만약 사용자의 진행 중인 스터디가 없다면, 코드에서 조건문 분기를 나누어 '진행 중인 스터디가 없습니다'를 대신 출력)
2. 사용자가 ```더 보기``` 버튼을 클릭하면, ```public/js/mypage.js``` 에 정의된 ```fetchLeadGroup()``` 함수를 호출한다. 함수 내부에서 fetch API는 쿼리스트링에 파라미터를 추가한 ```/mypage/fetch-leadgroup?load=count```로 요청을 보낸다. count 변수는 더 보기 버튼이 클릭될 때마다 +4씩 늘어난다. (4개씩 데이터를 추가하기 위해)
3. fetch 요청을 받은 서버 사이드(```routes/mypage.js```에서는 쿼리 스트링 파라미터로 전달받은 숫자를 해석한다. 해당 숫자는 DB에서 데이터를 SELECT할 시작 인덱스(```startIdx```)이다. startIdx에 4를 더한 ```endIdx```를 함께 활용하여 DB에 4개의 추가 데이터를 요청하는 query를 보낸다.
4. DB로부터 가져온 추가 데이터를 json 포맷으로 클라이언트 사이드에 응답해준다. (```res.json```)
5. 클라이언트 사이드의 js 파일(```public/js/mypage.js```)에서는 fetch를 통해 받아온 데이터를 후속처리(```.then```) 하여 파싱하고, 파싱된 각 데이터를 ```createLeadgroup()``` 함수를 사용하여 화면에 그려준다. createLeadgroup() 함수는 ```createElement``` 메서드를 사용하여 lead-group-list의 ```group-list__item```요소들을 생성하는 함수이다.
6. ```더 보기``` 버튼을 클릭할 때마다 쿼리 스트링에 전달되는 ```count``` 변수가 4씩 늘어나기 때문에 새로운 데이터는 순서대로 4개씩 추가된다. 다른 페이지를 방문했다가 다시 마이페이지로 접속하는 경우 count는 초기화된다.



### 작동 테스트
![mypage_ssr+csr](https://user-images.githubusercontent.com/75300807/126867441-0921b590-75a4-4aab-9fa4-bf22f279e7de.gif)

youngToMaturity committed:
feat: mypage 로드시, 로그인 된 user가 가입/생성한 그룹의 데이터를 fetch

각 스터디 클릭시 group-edit에 group_id를 넘길 수 있도록 임시로 실행하고자 하는 과정. (mypage가 완성되면 수정 예정)

youngToMaturity committed:
feat: fetch를 통해 받은 study_gorup_id를 각 group-list__edit-btn의 value로 지정

[임시]
! issue : 각 lead-group-list의 group-list__item의 갯수를 어떻게 지정할 것인가? - user마다 운영하고 있는 스터디 그룹의 갯수가 다를것이기 때문에
! issue: group-list__item에 해당 study_group_id까지는 지정할 수 있겠지만, 해당 div의 내용들을 어떻게 바꿀 것인가? (fetch속의 fetch?)

youngToMaturity committed:
feat : 관리 버튼 클릭시 각 item에 해당하는 study-group-id post

관리 버튼 클릭시 group-list-hidden__input에 저장된 study-group-id가 post된 후 group-edit이 render되도록 함

youngToMaturity committed:
feat: '관리' 버튼 클릭시 post로 group-edit.ejs에 선택된 그룹의 정보 넘기기

db.query를 활용하여 submit된 id와 같은 study_group을 찾고, 해당 그룹에 속해있는 group_member도 result로 저장. 하지만 ejs에 대한 이해도 부족으로 이 값들을 ejs 상에 표시하는것 미완성

42KIM committed:
feat: Load lead-group list by ssr+csr

youngToMaturity commented on views/includes/mypage/lead-group-list.ejs:
```diff
@@ -1,37 +1,35 @@
-    <div class="group-tag__cat-sub">중분류</div>
-    <div class="group-tag__loc">장소</div>
-    <div class="group-tag__gndr">성별</div>
+    <div class="group-tag__cat-main"><%= data.main_category %></div>
```
<div> 태그 안에 이렇게 넣을 수 있군요 참고해서 다른 페이지도 작성하겠습니다!

youngToMaturity committed:
style :  ejs 적용과정 공유를 위한 커밋

youngToMaturity committed:
feat : group-edit 페이지에서 기존의 스터디 그룹 정보를 기본값으로 표시해주기

수정하기 페이지에 들어가는 경우, 해당 스터디의 기존 정보를 page에 보이도록 하였습니다. ( 중분류, 멤버 미구현 )

youngToMaturity committed:
feat: group-edit 페이지 상의 중분류 select를 수정하고자 하는 스터디 그룹의 기존 값으로 지정

route/mypage.js의 /fetch를 /fetch-group으로 수정
js/group-edit.js의 categoryChange(e)에 있던 중분류 list를 전역으로 선언하여 DOMContentLoad시와 change시 모두 활용할 수 있도록 함.
categotyChange(e) 함수내부의 중분류 Element들의 생성과 지정을 setSub()함수로 처리하였고, hidden Input을 통해 확인할 수 있는 render시에 받은 sub과 같은 경우 selected 옵션을 추가.

42KIM committed:
refactor: Change MySQL BETWEEN to LIMIT

42KIM committed:
style: Change variable name

42KIM commented:
## 추가 내용

### study_group 테이블에 create_date DATETIME, NOT NULL 추가
+ 데이터 정렬 목적의 create_date 컬럼 추가. 스터디 생성 시 NOW() 값 추가
+ idx 컬럼을 삭제. 기존 id 컬럼을 PRIMARY KEY로 복귀

### DB 순차 접근을 위해 BETWEEN 문을 LIMIT 문으로 변경
+ 사용법 : LIMIT [시작 행의 인덱스], [가져올 행 개수] 
+ 테이블 인덱스는 0부터 시작
+ 사용 예시 : ```ORDER BY create_date LIMIT [시작 인덱스], [개수(4개)]```

### 변수명 수정
+ 쿼리 스트링의 ```count```를 ```idx```로 수정 (LIMIT의 시작 인덱스를 의미하기 때문)
+ fetchLeadGroup -> getGroups
+ createLeadGroup -> createGroup

### 🚨서브 쿼리 문 추가🚨
마이페이지에 접속했을 때, 유저 별로 다른 참여 중인 스터디/진행 중인 스터디 리스트를 출력해주기 위해, 현재 로그인된 사용자를 식별하는 것이 필요하다.
1. 현재 로그인된 사용자가 가입한 스터디 id를 찾아서
2. 그 결과를 본인이 스터디장인 경우(is_manager=1)와 아닌 경우(is_manager=0)로 구분한다.

이를 위해,
1. **현재 유저의 id를 가지고 group_member 테이블에 접근하여 study_group_id를 얻은 뒤
2. 취득한 study_group_id를 가지고 다시 study_group 테이블에 접근하여 스터디 정보를 가져온다.

의논 해볼 부분이 좀 있음...
- JOIN? 서브쿼리?
- 어느 테이블에서 정렬을 하고, LIMIT을 걸 건지

youngToMaturity committed:
feat: 멤버의 수만큼 div를 생성하여 닉네임과 강퇴버튼을 부여

강퇴 버튼의 value에 group-member의 id를 지정했고, 닉네임은 각 회원의 닉네임으로 지정했다. 현재는 임시로 방장도 출력했지만 추후에는 방장을 제외하고 출력하도록 처리할 것

42KIM committed:
feat: 마이페이지 다중쿼리, 서브쿼리 추가

42KIM commented:
## 추추가 내용

### '참여 중인 스터디' 리스트 출력 구현
+ 단, 현재 스터디 생성하기 만 존재하고 스터디 참여하기는 구현이 되지 않은 상태이므로 더미 데이터를 직접 만들어 테스트 해야함
+ 임시 '더 보기' 버튼 추가

### group_member 테이블에 create_date 추가
+ 데이터를 정렬하여 4개씩 가져오기 위해 정렬 목적의 컬럼 생성

### 마이페이지 쿼리문 추가
+ 다중 쿼리 : 마이페이지 접속 시, 참여 중인 스터디/진행 중인 스터디 리스트를 모두 가져오기 위해 is_manager가 1인 경우와 0인 경우를 나눠 쿼리 전송
+ 서브 쿼리 : 다중 쿼리의 각각은 study_group에서 스터디 정보까지 가져오기 위해 서브쿼리 보유
+ 가져온 데이터가 담긴 배열의 [0]은 leadgroups, [1]은 joingroups. 각각 lead-group-list.ejs와 join-group-list.ejs로 전달 

### 무한 스크롤 기능 추가
+ ```더 보기``` 버튼 삭제
+ ```scrollHeight```, ```clientHeight```, ```scrollTop``` 프로퍼티 활용

42KIM committed:
feat: 무한스크롤 기능 구현, 더보기 버튼 삭제

youngToMaturity committed:
feat : 수정완료 버튼 클릭시 db 수정, 예외처리 두가지 적용

만약 db상에 수정하고자 하는 이름이 있고 && 그 이름이 현재 수정하고자 하는 스터디 그룹의 이름이 아니라면 수정불가.
만약 현재 스터디 인원수보다 수정하고자 하는 스터디 정원이 더 적다면 수정불가.
그 외의 상황은 수정 가능

youngToMaturity committed:
feat: 강퇴 버튼 클릭시 해당 멤버 강퇴 기능 구현

group-edit.ejs - form 위치 수정 (css 재지정 필요)
routes/mypage.js - kickoutForm post 처리, 방장인 경우(추후 방장은 members에 보이지 않도록 처리 요망) 강퇴안됨
js/mypage.js - Kickout시 방생되는 status 2개 추가, 기존의 complete은 edit으로 수정 ?: 각 case 처리 이후에 redirect할 수 있을까?
js/group-edit.js 강퇴 버튼 클릭시 kickout 폼 submit

youngToMaturity reviewed (approved):
현재 코드상에서 더보기는 각 `참여 중인 스터디`와 `진행 중인 스터디` 항목의 가장 아래까지 스크롤 하는 경우겠군요! 75% 정도 됐을때 두번 불러오는 사항은 issue에 올려두고 추후에 처리하면 좋을 것 같습니다 수고하셨습니다

youngToMaturity committed:
Merge remote-tracking branch 'origin/ssr+csr' into group-edit

# Conflicts:
#	public/js/mypage.js
#	routes/mypage.js
#	views/includes/mypage/lead-group-list.ejs
